### PR TITLE
[babel 8] Remove `getChalk` function

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

There is no good reason for us to have chalk exposed as part of our public API, and it's making the upgrade to v4/5 incredibly complex.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15812"></a>

### CODE REVIEW

**Comment by JLHwung** [comment]
I guess `getChalk` was there because the code-frame test wants to manipulate the Chalk instance used by the highlight package. If this API is removed we should resolve chalk from the highlight package.

**Comment by nicolo-ribaudo** [comment]
Yeah thanks, I was just reading https://github.com/babel/babel/pull/15499#discussion_r1144203458 some minutes ago :)

**Comment by nicolo-ribaudo** [comment]
image

Oh no, we miss chalk 3!

File: packages/babel-highlight/test/index.js [hunk 1]
```diff
@@ -86,7 +92,12 @@ describe("@babel/highlight", function () {
     });
   });
 
-  describe("getChalk", function () {
+  describeBabel7NoESM("getChalk", function () {
+    let getChalk;
+    beforeAll(function () {
+      getChalk = require("../lib/index.js");
```
Comment (nicolo-ribaudo) [inline packages/babel-highlight/test/index.js:98]:
```suggestion
      getChalk = require("../lib/index.js").getChalk;
```

File: packages/babel-highlight/test/index.js [hunk 2]
```diff
@@ -1,22 +1,30 @@
+import { USE_ESM } from "$repo-utils";
+
 import stripAnsi from "strip-ansi";
+import chalk from "chalk";
+
+import { createRequire } from "module";
```
Comment (nicolo-ribaudo) [inline packages/babel-highlight/test/index.js:6]:
```suggestion
import { createRequire } from "module";
const require = createRequire(import.meta.url);
```

File: packages/babel-highlight/test/index.js [hunk 3]
```diff
@@ -86,18 +95,22 @@ describe("@babel/highlight", function () {
     });
   });
 
-  describe("getChalk", function () {
+  describeBabel7NoESM("getChalk", function () {
+    const { getChalk } = require("../lib/index.js");
```
Comment (nicolo-ribaudo) [inline packages/babel-highlight/test/index.js:99]:
```suggestion
	let getChalk;
	beforeAll(() => {
      ({ getChalk } = require("../lib/index.js"));
    });
```

**Comment by nicolo-ribaudo** [comment]
CI is finally green

File: packages/babel-highlight/test/index.js [hunk 4]
```diff
@@ -86,18 +95,25 @@ describe("@babel/highlight", function () {
...
+	beforeAll(() => {
+      ({ getChalk } = require("../lib/index.js"));
+    });
+
     describe("when colors are supported", function () {
       stubColorSupport(true);
 
       describe("when forceColor is not passed", function () {
         it("returns a Chalk instance", function () {
+          const { getChalk } = require("../lib/index.js");
```
Comment (nicolo-ribaudo) [inline packages/babel-highlight/test/index.js:109]: Remove these lines

File: packages/babel-highlight/test/index.js [hunk 5]
```diff
@@ -1,22 +1,31 @@
...
+
+    beforeEach(async function () {
       originalChalkSupportsColor = chalk.supportsColor;
       originalChalkLevel = chalk.level;
       originalChalkEnabled = chalk.enabled;
+
       chalk.supportsColor = supported ? { level: 1 } : false;
       chalk.level = supported ? 1 : 0;
-      chalk.enabled = supported;
+      chalk.enabled = true;
```
Comment (nicolo-ribaudo) [inline packages/babel-highlight/test/index.js:28]: Revert changes to this function

**Review by liuxingbaoyu — APPROVED** [review]

File: packages/babel-highlight/src/index.ts [hunk 6]
```diff
@@ -249,25 +247,30 @@ type Options = {
...
- * The Chalk instance that should be used given the passed options.
- */
-export function getChalk(options: Options) {
-  return options.forceColor
-    ? new Chalk.constructor({ enabled: true, level: 1 })
-    : Chalk;
+let chalkWithForcedColor: Chalk = undefined;
+function getChalk(forceColor: boolean) {
+  if (forceColor) {
+    chalkWithForcedColor ??= new chalk.constructor({ enabled: true, level: 1 });
```
Comment (JLHwung) [inline packages/babel-highlight/src/index.ts:256]: If `forceColor` is `true`, `getChalk` will return a memoized chalk instance, previously it returned a new instance. The new behaviour aligns with the one when `forceColor` is `false`.

**Review by JLHwung — APPROVED** [review]

**Comment by JLHwung** [comment]
Can you update https://github.com/babel/website/pull/2765?

File: packages/babel-highlight/src/index.ts (hunk 6 above)
Comment (nicolo-ribaudo) [reply packages/babel-highlight/src/index.ts:256]: Yeah I noticed it while changing this -- I think it's ok, it has the benefit of allocating less objects (+ as you noted aligning with `forceColor: false`) and no actual disadvantages.

**Pull request merged**